source/server/overload_manager_impl.cc

```diff
@@ -18,22 +19,55 @@ namespace Server {
 
 namespace {
 
-class ThresholdTriggerImpl : public OverloadAction::Trigger {
+class ThresholdTriggerImpl final : public OverloadAction::Trigger {
 public:
   ThresholdTriggerImpl(const envoy::config::overload::v3::ThresholdTrigger& config)
-      : threshold_(config.value()) {}
+      : threshold_(config.value()), state_(OverloadActionState::inactive()) {}
 
   bool updateValue(double value) override {
-    const bool fired = isFired();
-    value_ = value;
-    return fired != isFired();
+    const OverloadActionState state = actionState();
+    state_ =
+        value >= threshold_ ? OverloadActionState::saturated() : OverloadActionState::inactive();
+    return state.value() != actionState().value();
   }
 
-  bool isFired() const override { return value_.has_value() && value_ >= threshold_; }
+  OverloadActionState actionState() const override { return state_; }
 
 private:
   const double threshold_;
-  absl::optional<double> value_;
+  OverloadActionState state_;
+};
+
+class ScaledTriggerImpl final : public OverloadAction::Trigger {
+public:
+  ScaledTriggerImpl(const envoy::config::overload::v3::ScaledTrigger& config)
+      : scaling_threshold_(config.scaling_threshold()),
+        saturated_threshold_(config.saturation_threshold()),
+        state_(OverloadActionState::inactive()) {
+    if (scaling_threshold_ >= saturated_threshold_) {
+      throw EnvoyException("min_value must be less than max_value");
```

Comment:
Nit: can you adjust this exception message to match the field names? I.e. scaling/saturated_threshold.